tuist/tuist
# Added proper support for merged non-runnable binaries

## Merged non-runnable binaries support

**TL;DR**: Fixed frameworks and xcframeworks being embedded into the app after being merged to a non-runnable target (e.g., another dynamic framework).

### Short description 📝

Tuist supports mergeable frameworks introduced in Xcode 15.0 only for executables (apps, app extensions). The main purpose of tuist for mergeable support is to ensure that it doesn't embed merged frameworks separately into the app, preventing duplication.

However, if a separate framework is created to share the same dependencies between separate targets, Tuist will add merged local frameworks and xcframeworks to the list of embedded frameworks.

This MR fixes the issue for both frameworks and xcframeworks by inspecting the main target dependencies in the `embeddableFrameworks` phase of `GraphTraversal`. If any of them is a merged binary, its mergeable dependencies are excluded from the embedded list.

### How to test the changes locally 🧐

- Generate Project for `ios_app_with_shared_mergeable_framework` fixture
- Open Run Phases -> Embedded frameworks for the main App.
- Ensure only `SharedDependencies.framework` is embedded

### Results
| Before | After |
| ------ | ------ |
| <img width="420" alt="image" src="https://github.com/tuist/tuist/assets/31652265/b4b49dcb-3b3c-4dc4-a6d6-0038375c634b"> | <img width="420" alt="image" src="https://github.com/tuist/tuist/assets/31652265/d6883714-5c62-4469-83eb-4a07ce90f1f4"> |

### Notes and potential improvements for the future

If a merged framework with mergeable dependencies is created, it has to be linked to the main targets instead of those dependencies that were merged. An extra linting phase could probably be added to ensure that behavior. 

### Contributor checklist ✅

- [x] The change is tested via unit testing, acceptance testing, or both

### Reviewer checklist ✅
